crates/bevy_ecs/src/schedule/system_graph.rs

```diff
@@ -0,0 +1,264 @@
+/// `SystemGraphJoinExt::join_into` is implemented on any type that iterates over
+/// `SystemGraphNode`.
+/// ```
+/// # use bevy_ecs::prelude::*;
+/// # fn sys_a() {}
+/// # fn sys_b() {}
+/// # fn sys_c() {}
+/// # fn sys_d() {}
+/// # fn sys_e() {}
+/// let graph = SystemGraph::new();
+///
+/// let start_a = graph.root(sys_a.system());
+/// let start_b = graph.root(sys_b.system());
+/// let start_c = graph.root(sys_c.system());
+///
+/// vec![start_a, start_b, start_c]
+///     .join_into(sys_d.system())
+///     .then(sys_e.system());
+///
+/// // Convert into a SystemSet
+/// let system_set: SystemSet = graph.into();
+/// ```
+#[derive(Clone, Default)]
+pub struct SystemGraph(Arc<Mutex<HashMap<NodeId, SystemDescriptor>>>);
+
+impl SystemGraph {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn root(&self, system: impl Into<SystemDescriptor>) -> SystemGraphNode {
+        self.create_node(system.into())
+    }
+
+    pub fn is_same_graph(&self, other: &Self) -> bool {
+        Arc::ptr_eq(&self.0, &other.0)
+    }
+
+    fn create_node(&self, mut system: SystemDescriptor) -> SystemGraphNode {
+        let id = NodeId(NEXT_NODE_ID.fetch_add(1, Ordering::Relaxed));
+        match &mut system {
+            SystemDescriptor::Parallel(descriptor) => descriptor.labels.push(id.dyn_clone()),
+            SystemDescriptor::Exclusive(descriptor) => descriptor.labels.push(id.dyn_clone()),
+        }
+        self.0.lock().unwrap().insert(id, system);
+        SystemGraphNode {
+            id,
+            graph: self.clone(),
+        }
+    }
+
+    fn add_dependency(&self, src: NodeId, dst: NodeId) {
+        if let Some(system) = self.0.lock().unwrap().get_mut(&dst) {
+            match system {
+                SystemDescriptor::Parallel(descriptor) => descriptor.after.push(src.dyn_clone()),
+                SystemDescriptor::Exclusive(descriptor) => descriptor.after.push(src.dyn_clone()),
+            }
+        } else {
+            panic!(
+                "Attempted to add dependency for {:?}, which doesn't exist.",
+                dst
+            );
+        }
+    }
+}
+
+/// A draining conversion from [SystemGraph] to [SystemSet]. All other clones of the same graph
+/// will be empty afterwards.
+///
+/// [SystemGraph]: crate::schedule::SystemSet
+/// [SystemSet]: crate::schedule::SystemSet
+impl From<SystemGraph> for SystemSet {
+    fn from(graph: SystemGraph) -> Self {
+        let mut system_set = SystemSet::new();
+        for (_, system) in graph.0.lock().unwrap().drain() {
+            system_set = system_set.with_system(system);
+        }
+        system_set
+    }
+}
+
+#[derive(Clone)]
+pub struct SystemGraphNode {
+    id: NodeId,
+    graph: SystemGraph,
+}
+
+impl SystemGraphNode {
+    pub fn then(&self, next: impl Into<SystemDescriptor>) -> SystemGraphNode {
+        let node = self.graph.create_node(next.into());
+        self.add_dependent(node.id);
```

Comment:
```suggestion
        self.graph.add_dependency(self.id, dst);
```
`add_dependent` is a single line function, it should probably just be inlined here.